File: ios/VitalView/Source/DataSource.swift
//
//  DataSource.swift
//  VitalView
//
//  The single interface that the UI talks to. Both the live BLE source and the mock
//  source implement it, so every screen works with or without hardware. Packets
//  delivered to consumers carry EXTENDED device timestamps (see Timebase.swift).
//
//  Mirrors web/src/source/DataSource.ts. Event delivery uses Combine subjects so the
//  AppModel can subscribe with backpressure-free sinks (the high-rate packet handlers
//  push into non-reactive ring buffers, not @Published state).
//

import Combine
import Foundation

/// Connection lifecycle, mirroring the web `ConnectionState` union.
enum ConnectionState: String {
    case disconnected
    case unsupported
    case scanning
    case connecting
    case connected
    case reconnecting
}

/// Optional detail attached to a connection-state change.
struct ConnectionDetail {
    var error: String?
}

/// Combine subjects a `DataSource` publishes on. The AppModel subscribes to these.
/// Subjects are `Void`-failing PassthroughSubjects (events only, no completion).
final class DataSourceEvents {
    let ecg = PassthroughSubject<EcgPacket, Never>()
    let bioz = PassthroughSubject<BiozPacket, Never>()
    let ppg = PassthroughSubject<PpgPacket, Never>()
    /// Firmware-computed metrics (optional; the app computes its own in parallel).
    let metrics = PassthroughSubject<MetricsPacket, Never>()
    let status = PassthroughSubject<StatusMessage, Never>()
    let info = PassthroughSubject<InfoMessage, Never>()
    let battery = PassthroughSubject<Int, Never>()
    /// (stream, newly dropped packet count) from seq gaps.
    let dropped = PassthroughSubject<(StreamKind, Int), Never>()
    let connection = PassthroughSubject<(ConnectionState, ConnectionDetail?), Never>()
}

/// The data-source abstraction implemented by `MockSource` and `BleSource`.
protocol DataSource: AnyObject {
    var isMock: Bool { get }
    var state: ConnectionState { get }
    /// offset = hostNowUs − deviceTus, or nil before SYNC_CLOCK.
    var clockOffsetUs: Double? { get }
    var deviceName: String? { get }

    /// The Combine event hub for this source.
    var events: DataSourceEvents { get }

    func connect()
    func disconnect()
    func start(streamMask: UInt8)
    func stop()
    func setRate(_ stream: StreamKind, hz: Int)
    func syncClock()
    func getInfo()
}

/// Shared base providing the event hub + connection-state plumbing.
class BaseDataSource: DataSource {
    let events = DataSourceEvents()
    private(set) var stateValue: ConnectionState = .disconnected

    var isMock: Bool { false }
    var clockOffsetUs: Double? { nil }
    var deviceName: String? { nil }

    var state: ConnectionState { stateValue }

    func setState(_ state: ConnectionState, detail: ConnectionDetail? = nil) {
        stateValue = state
        events.connection.send((state, detail))
    }

    // Subclasses override these.
    func connect() {}
    func disconnect() {}
    func start(streamMask: UInt8) {}
    func stop() {}
    func setRate(_ stream: StreamKind, hz: Int) {}
    func syncClock() {}
    func getInfo() {}
}

File: web/src/source/DataSource.ts
/**
 * The single interface that the UI talks to. Both the live BLE source and the mock
 * source implement it, so every screen works with or without hardware (§8). Packets
 * delivered to consumers carry EXTENDED device timestamps (see timebase.ts).
 */
import type {
  BiozPacket,
  EcgPacket,
  InfoMessage,
  MetricsPacket,
  PpgPacket,
  StatusMessage,
} from '../ble/packets';
import type { StreamKind } from '../ble/protocol';
import { Emitter } from '../util/emitter';

export type ConnectionState =
  | 'disconnected'
  | 'unsupported'
  | 'scanning'
  | 'connecting'
  | 'connected'
  | 'reconnecting';

export interface DataSourceEvents {
  ecg: (p: EcgPacket) => void;
  bioz: (p: BiozPacket) => void;
  ppg: (p: PpgPacket) => void;
  /** Firmware-computed metrics (optional; the app computes its own in parallel). */
  metrics: (p: MetricsPacket) => void;
  status: (s: StatusMessage) => void;
  info: (i: InfoMessage) => void;
  battery: (pct: number) => void;
  /** Newly dropped packet count for a stream (from seq gaps). */
  dropped: (stream: StreamKind, count: number) => void;
  connection: (state: ConnectionState, detail?: { error?: string }) => void;
}

export interface DataSource {
  readonly isMock: boolean;
  readonly state: ConnectionState;
  /** offset = hostNowUs − deviceTus, or null before SYNC_CLOCK. */
  readonly clockOffsetUs: number | null;
  readonly deviceName: string | null;

  connect(): Promise<void>;
  disconnect(): Promise<void>;
  start(streamMask: number): Promise<void>;
  stop(): Promise<void>;
  setRate(stream: StreamKind, hz: number): Promise<void>;
  syncClock(): Promise<void>;
  getInfo(): Promise<void>;

  on<K extends keyof DataSourceEvents>(event: K, cb: DataSourceEvents[K]): () => void;
}

/** Shared base providing the emitter plumbing. */
export abstract class BaseDataSource implements DataSource {
  protected readonly emitter = new Emitter<DataSourceEvents>();
  protected _state: ConnectionState = 'disconnected';

  abstract readonly isMock: boolean;
  abstract clockOffsetUs: number | null;
  abstract deviceName: string | null;

  get state(): ConnectionState {
    return this._state;
  }

  protected setState(state: ConnectionState, detail?: { error?: string }): void {
    this._state = state;
    this.emitter.emit('connection', state, detail);
  }

  on<K extends keyof DataSourceEvents>(event: K, cb: DataSourceEvents[K]): () => void {
    return this.emitter.on(event, cb);
  }

  abstract connect(): Promise<void>;
  abstract disconnect(): Promise<void>;
  abstract start(streamMask: number): Promise<void>;
  abstract stop(): Promise<void>;
  abstract setRate(stream: StreamKind, hz: number): Promise<void>;
  abstract syncClock(): Promise<void>;
  abstract getInfo(): Promise<void>;
}
